flash/data/data_pipeline.py

```diff
@@ -0,0 +1,612 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import functools
+import weakref
+from typing import Any, Callable, Iterable, Optional, Sequence, Tuple, Type, TYPE_CHECKING, Union
+
+from pytorch_lightning.trainer.connectors.data_connector import _PatchDataLoader
+from pytorch_lightning.trainer.states import RunningStage
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch.utils.data._utils.collate import default_collate, default_convert
+from torch.utils.data.dataloader import DataLoader
+
+from flash.data.auto_dataset import AutoDataset
+from flash.data.batch import _PostProcessor, _PreProcessor, _Sequential
+from flash.data.process import Postprocess, Preprocess
+from flash.data.utils import _STAGES_PREFIX
+
+if TYPE_CHECKING:
+    from flash.core.model import Task
+
+
+class DataPipeline:
+    """
+    DataPipeline handles the connecting logic between ``Preprocess``, ``PostProcess``,
+    ``DataModule``, and ``LightningModule`` depending on the current ``RunningStage``
+
+    The ``Preprocess`` hooks are used to generate several objects:
+
+    1. Generate an ``AutoDataset`` from ``load_data`` and ``load_sample``.
+
+        Example::
+            class AutoDataset
+                def __init__(self, ..., data, ...):
+                    self.preprocessed_data: Iterable = Preprocess.load_data(data)
+
+                def __getitem__(self, index):
+                    return Preprocess.load_sample(self.preprocessed_data[index])
+
+                def __len__(self):
+                    return len(self.preprocessed_data)
+
+    2. Create a ``worker_collate_fn`` which is injected directly into the ``DataLoader``
+       and a ``device_collate_fn`` injected after ``LightningModule.transfer_batch_to_device`` hook.
+
+        Objects description:
+
+        _Sequential:
+            ┌────────────────────────────────────┐
+            │  per_sample_pre_tensor_transform   │
+            │                |                   │
+            │  per_sample_to_tensor_transform    │
+            │                |                   │
+            │  per_sample_post_tensor_transform  │
+            └────────────────────────────────────┘
+
+        _PreProcessor:
+
+            The ``_PreProcessor`` performs ``per_sample_transform``, ``collate``, ``per_batch_transform`` as follow:
+
+            ``per_batch_transform`` and ``per_sample_transform_on_device`` are mutually exclusive
+
+            def forward(self, samples: Sequence[Any]):
+                    samples = [self.per_sample_transform(sample) for sample in samples]
+                    samples = type(samples)(samples)
+                    samples = self.collate_fn(samples)
+                samples = self.per_batch_transform(samples)
+                return samples
```

Comment:
add left space as this probably won't  be correctly rendered as code or add explicitly code tag